src/libraries/System.IO.Compression/src/System/IO/Compression/ZipBlocks.Sync.cs

```diff
@@ -0,0 +1,330 @@
+            GeneralPurposeBitFlag = BinaryPrimitives.ReadUInt16LittleEndian(buffer[FieldLocations.GeneralPurposeBitFlags..]),
+            CompressionMethod = BinaryPrimitives.ReadUInt16LittleEndian(buffer[FieldLocations.CompressionMethod..]),
+            LastModified = BinaryPrimitives.ReadUInt32LittleEndian(buffer[FieldLocations.LastModified..]),
+            Crc32 = BinaryPrimitives.ReadUInt32LittleEndian(buffer[FieldLocations.Crc32..]),
+            FilenameLength = BinaryPrimitives.ReadUInt16LittleEndian(buffer[FieldLocations.FilenameLength..]),
+            ExtraFieldLength = BinaryPrimitives.ReadUInt16LittleEndian(buffer[FieldLocations.ExtraFieldLength..]),
+            FileCommentLength = BinaryPrimitives.ReadUInt16LittleEndian(buffer[FieldLocations.FileCommentLength..]),
+            InternalFileAttributes = BinaryPrimitives.ReadUInt16LittleEndian(buffer[FieldLocations.InternalFileAttributes..]),
+            ExternalFileAttributes = BinaryPrimitives.ReadUInt32LittleEndian(buffer[FieldLocations.ExternalFileAttributes..])
+        };
+
+        uint compressedSizeSmall = BinaryPrimitives.ReadUInt32LittleEndian(buffer[FieldLocations.CompressedSize..]);
+        uint uncompressedSizeSmall = BinaryPrimitives.ReadUInt32LittleEndian(buffer[FieldLocations.UncompressedSize..]);
+        ushort diskNumberStartSmall = BinaryPrimitives.ReadUInt16LittleEndian(buffer[FieldLocations.DiskNumberStart..]);
+        uint relativeOffsetOfLocalHeaderSmall = BinaryPrimitives.ReadUInt32LittleEndian(buffer[FieldLocations.RelativeOffsetOfLocalHeader..]);
+
+        // Assemble the dynamic header in a separate buffer. We can't guarantee that it's all in the input buffer,
+        // some additional data might need to come from the stream.
+        int dynamicHeaderSize = header.FilenameLength + header.ExtraFieldLength + header.FileCommentLength;
+        int remainingBufferLength = buffer.Length - FieldLocations.DynamicData;
+        int bytesToRead = dynamicHeaderSize - remainingBufferLength;
+        scoped ReadOnlySpan<byte> dynamicHeader;
+        byte[]? arrayPoolBuffer = null;
+
+        Zip64ExtraField zip64;
+
+        try
+        {
+            // No need to read extra data from the stream, no need to allocate a new buffer.
+            if (bytesToRead <= 0)
+            {
+                dynamicHeader = buffer[FieldLocations.DynamicData..];
+            }
+            // Data needs to come from two sources, and we must thus copy data into a single address space.
+            else
+            {
+                if (dynamicHeaderSize > StackAllocationThreshold)
+                {
+                    arrayPoolBuffer = ArrayPool<byte>.Shared.Rent(dynamicHeaderSize);
+                }
+
+                Span<byte> collatedHeader = dynamicHeaderSize <= StackAllocationThreshold ? stackalloc byte[StackAllocationThreshold].Slice(0, dynamicHeaderSize) : arrayPoolBuffer.AsSpan(0, dynamicHeaderSize);
```

Comment:
Is the compiler smart enough to not produce a warning for `arrayPoolBuffer` being null here?

```suggestion
                Span<byte> collatedHeader = dynamicHeaderSize <= StackAllocationThreshold
                    ? stackalloc byte[StackAllocationThreshold]
                    : (arrayPoolBuffer = ArrayPool<byte>.Shared.Rent(dynamicHeaderSize));
                collatedHeader = collatedHeader.Slice(0, dynamicHeaderSize);
```

Reply:
I wasn't seeing any warnings. I rewrote this part btw.